File: implementations/rust/src/canonicalize_mzml.rs
//! Canonical content hashing for mzML files (v0, frozen).
//!
//! Independent port of `canonicalize_mzml.py`. Hashes spectra only; every
//! `binaryDataArray` contributes a labeled (role, precision, count,
//! sha256-hex) record, so tampering with e.g. an ion-mobility array
//! changes the hash. See `spec/canonicalization-mzml-v0.md`.

use base64::Engine;
use flate2::read::ZlibDecoder;
use roxmltree::{Document, Node};
use sha2::{Digest, Sha256};
use std::io::Read;
use std::path::Path;

use crate::errors::{ProvenanceError, Result};

const US: u8 = 0x1f;
const RS: u8 = 0x1e;
const MZML_NS: &str = "http://psi.hupo.org/ms/mzml";
const MZML_CONTENT_DOMAIN: &[u8] = b"timsim.mzml.v0\x1f";

// PSI-MS accessions we resolve by number.
const ACC_MS_LEVEL: &str = "MS:1000511";
const ACC_SCAN_START_TIME: &str = "MS:1000016";
const ACC_POSITIVE: &str = "MS:1000130";
const ACC_NEGATIVE: &str = "MS:1000129";
const ACC_SELECTED_ION_MZ: &str = "MS:1000744";
const ACC_CHARGE_STATE: &str = "MS:1000041";
const ACC_ISO_TARGET: &str = "MS:1000827";
const ACC_ISO_LOWER: &str = "MS:1000828";
const ACC_ISO_UPPER: &str = "MS:1000829";
const ACC_F64: &str = "MS:1000523";
const ACC_F32: &str = "MS:1000521";
const ACC_I64: &str = "MS:1000522";
const ACC_I32: &str = "MS:1000519";
const ACC_NO_COMPRESSION: &str = "MS:1000576";
const ACC_ZLIB: &str = "MS:1000574";
const ACC_NUMPRESS_LINEAR: &str = "MS:1002312";
const ACC_NUMPRESS_PIC: &str = "MS:1002313";
const ACC_NUMPRESS_SLOF: &str = "MS:1002314";
const ACC_ION_MOBILITY: &str = "MS:1002476";
const ACC_ION_MOBILITY_ALT: &str = "MS:1003006";
const UNIT_MINUTE: &str = "UO:0000031";

const KNOWN_ARRAY_ROLE_ACCS: &[&str] = &[
    "MS:1000514", "MS:1000515", "MS:1000516", "MS:1000517", "MS:1000595",
    "MS:1000617", "MS:1000786", "MS:1000820", "MS:1000821", "MS:1000822",
    "MS:1002476", "MS:1002477", "MS:1002478", "MS:1003006", "MS:1003007",
    "MS:1003008", "MS:1003153",
];

const KNOWN_ENCODING_ACCS: &[&str] = &[
    "MS:1000519", "MS:1000521", "MS:1000522", "MS:1000523",
    "MS:1000574", "MS:1000576",
    "MS:1002312", "MS:1002313", "MS:1002314",
];

fn is_ms_elem(node: &Node, local: &str) -> bool {
    node.is_element() && node.tag_name().name() == local && node.tag_name().namespace() == Some(MZML_NS)
}

fn child_elem<'a, 'input>(parent: &Node<'a, 'input>, local: &str) -> Option<Node<'a, 'input>> {
    parent.children().find(|c| is_ms_elem(c, local))
}

fn children_elem<'a, 'input: 'a>(
    parent: &Node<'a, 'input>,
    local: &'a str,
) -> impl Iterator<Item = Node<'a, 'input>> + 'a {
    parent.children().filter(move |c| is_ms_elem(c, local))
}

fn cv_value<'a>(parent: &Node<'a, '_>, accession: &str) -> Option<&'a str> {
    for cv in children_elem(parent, "cvParam") {
        if cv.attribute("accession") == Some(accession) {
            return cv.attribute("value");
        }
    }
    None
}

fn cv_present(parent: &Node, accession: &str) -> bool {
    children_elem(parent, "cvParam").any(|cv| cv.attribute("accession") == Some(accession))
}

fn cv_value_recursive<'a>(root: &Node<'a, '_>, accession: &str) -> Option<&'a str> {
    for d in root.descendants() {
        if is_ms_elem(&d, "cvParam") && d.attribute("accession") == Some(accession) {
            return d.attribute("value");
        }
    }
    None
}

fn ieee754_hex(v: f64) -> String {
    hex::encode(v.to_be_bytes())
}

fn extract_polarity(spectrum: &Node) -> u8 {
    if cv_present(spectrum, ACC_POSITIVE) {
        return b'+';
    }
    if cv_present(spectrum, ACC_NEGATIVE) {
        return b'-';
    }
    b'?'
}

fn extract_rt_seconds(spectrum: &Node) -> Option<f64> {
    let scan_list = child_elem(spectrum, "scanList")?;
    let scan = child_elem(&scan_list, "scan")?;
    for cv in children_elem(&scan, "cvParam") {
        if cv.attribute("accession") == Some(ACC_SCAN_START_TIME) {
            let v = cv.attribute("value")?.parse::<f64>().ok()?;
            let is_minute = cv.attribute("unitAccession") == Some(UNIT_MINUTE);
            return Some(if is_minute { v * 60.0 } else { v });
        }
    }
    None
}

fn extract_ion_mobility(spectrum: &Node) -> Option<f64> {
    for acc in [ACC_ION_MOBILITY, ACC_ION_MOBILITY_ALT] {
        if let Some(v) = cv_value_recursive(spectrum, acc) {
            if let Ok(f) = v.parse::<f64>() {
                return Some(f);
            }
        }
    }
    None
}

#[derive(Default)]
struct PrecursorFields {
    target: Option<f64>,
    lower: Option<f64>,
    upper: Option<f64>,
    selected_mz: Option<f64>,
    charge: Option<i64>,
}

fn extract_precursor(spectrum: &Node) -> Option<PrecursorFields> {
    let plist = child_elem(spectrum, "precursorList")?;
    let precursor = child_elem(&plist, "precursor")?;

    let mut out = PrecursorFields::default();
    let mut any = false;

    if let Some(iso) = child_elem(&precursor, "isolationWindow") {
        for (slot, acc) in [
            ("target", ACC_ISO_TARGET),
            ("lower", ACC_ISO_LOWER),
            ("upper", ACC_ISO_UPPER),
        ] {
            if let Some(v) = cv_value(&iso, acc) {
                if let Ok(f) = v.parse::<f64>() {
                    any = true;
                    match slot {
                        "target" => out.target = Some(f),
                        "lower" => out.lower = Some(f),
                        "upper" => out.upper = Some(f),
                        _ => unreachable!(),
                    }
                }
            }
        }
    }

    if let Some(sil) = child_elem(&precursor, "selectedIonList") {
        if let Some(sion) = child_elem(&sil, "selectedIon") {
            if let Some(v) = cv_value(&sion, ACC_SELECTED_ION_MZ) {
                if let Ok(f) = v.parse::<f64>() {
                    any = true;
                    out.selected_mz = Some(f);
                }
            }
            if let Some(v) = cv_value(&sion, ACC_CHARGE_STATE) {
                if let Ok(i) = v.parse::<i64>() {
                    any = true;
                    out.charge = Some(i);
                }
            }
        }
    }

    if any {
        Some(out)
    } else {
        None
    }
}

fn decode_binary(bda: &Node) -> Result<Vec<u8>> {
    for acc in [ACC_NUMPRESS_LINEAR, ACC_NUMPRESS_PIC, ACC_NUMPRESS_SLOF] {
        if cv_present(bda, acc) {
            return Err(ProvenanceError::Canonicalization(format!(
                "mzml binaryDataArray uses numpress ({acc}); not supported in v0"
            )));
        }
    }
    let binary_el = child_elem(bda, "binary").ok_or_else(|| {
        ProvenanceError::MalformedSidecar(
            "mzml binaryDataArray is missing the inner <binary> element".into(),
        )
    })?;
    let text = binary_el.text().unwrap_or("").trim();
    if text.is_empty() {
        return Ok(Vec::new());
    }
    let raw = base64::engine::general_purpose::STANDARD
        .decode(text)
        .map_err(|e| {
            ProvenanceError::MalformedSidecar(format!("undecodable base64 in binary: {e}"))
        })?;
    if cv_present(bda, ACC_ZLIB) {
        let mut d = ZlibDecoder::new(&raw[..]);
        let mut out = Vec::new();
        d.read_to_end(&mut out).map_err(|e| {
            ProvenanceError::MalformedSidecar(format!("undecompressable zlib: {e}"))
        })?;
        return Ok(out);
    }
    // Explicit no-compression or missing cvParam — treat as raw bytes.
    let _ = ACC_NO_COMPRESSION;
    Ok(raw)
}

fn precision_tag(bda: &Node) -> (&'static [u8], usize) {
    if cv_present(bda, ACC_F64) {
        return (b"f64", 8);
    }
    if cv_present(bda, ACC_F32) {
        return (b"f32", 4);
    }
    if cv_present(bda, ACC_I64) {
        return (b"i64", 8);
    }
    if cv_present(bda, ACC_I32) {
        return (b"i32", 4);
    }
    (b"??", 0)
}

fn array_role_label(bda: &Node) -> Result<String> {
    for cv in children_elem(bda, "cvParam") {
        if let Some(acc) = cv.attribute("accession") {
            if KNOWN_ARRAY_ROLE_ACCS.contains(&acc) {
                return Ok(acc.to_owned());
            }
        }
    }
    for cv in children_elem(bda, "cvParam") {
        if let Some(acc) = cv.attribute("accession") {
            if !acc.is_empty() && !KNOWN_ENCODING_ACCS.contains(&acc) {
                return Ok(format!("unknown:{acc}"));
            }
        }
    }
    Err(ProvenanceError::MalformedSidecar(
        "binaryDataArray has no cvParam identifying its array role".into(),
    ))
}

fn emit(out: &mut Vec<u8>, key: &[u8], value: &[u8]) {
    out.push(US);
    out.extend_from_slice(key);
    out.push(US);
    out.extend_from_slice(value);
    out.push(US);
}

fn spectrum_record(spectrum: &Node) -> Result<Vec<u8>> {
    let spec_id = spectrum.attribute("id").unwrap_or("");
    let index_str = spectrum.attribute("index").unwrap_or("");
    let index: i64 = index_str.parse().map_err(|_| {
        ProvenanceError::MalformedSidecar(format!(
            "mzml spectrum {spec_id:?} has missing or non-integer index"
        ))
    })?;

    let ms_level = cv_value(spectrum, ACC_MS_LEVEL).unwrap_or("");
    let polarity = extract_polarity(spectrum);
    let rt = extract_rt_seconds(spectrum);
    let mob = extract_ion_mobility(spectrum);
    let precursor = extract_precursor(spectrum);

    // Collect every binary array; sort by role label.
    let mut arrays: Vec<(String, &'static [u8], usize, String)> = Vec::new();
    if let Some(bdal) = child_elem(spectrum, "binaryDataArrayList") {
        for bda in children_elem(&bdal, "binaryDataArray") {
            let role = array_role_label(&bda)?;
            let payload = decode_binary(&bda)?;
            let (tag, width) = precision_tag(&bda);
            let count = if width != 0 { payload.len() / width } else { 0 };
            let mut h = Sha256::new();
            h.update(&payload);
            let digest = hex::encode(h.finalize());
            arrays.push((role, tag, count, digest));
        }
    }
    arrays.sort_by(|a, b| a.0.cmp(&b.0));

    let mut rec = Vec::with_capacity(512);

    emit(&mut rec, b"spec_index", index.to_string().as_bytes());
    emit(&mut rec, b"spec_id", spec_id.as_bytes());
    emit(&mut rec, b"ms_level", ms_level.as_bytes());
    emit(&mut rec, b"polarity", &[polarity]);
    emit(
        &mut rec,
        b"rt_sec",
        rt.map(ieee754_hex).as_deref().unwrap_or("").as_bytes(),
    );
    emit(
        &mut rec,
        b"mobility",
        mob.map(ieee754_hex).as_deref().unwrap_or("").as_bytes(),
    );

    match &precursor {
        None => {
            emit(&mut rec, b"prec_target", b"");
            emit(&mut rec, b"prec_lower", b"");
            emit(&mut rec, b"prec_upper", b"");
            emit(&mut rec, b"prec_selected", b"");
            emit(&mut rec, b"prec_charge", b"");
        }
        Some(p) => {
            emit(
                &mut rec,
                b"prec_target",
                p.target.map(ieee754_hex).as_deref().unwrap_or("").as_bytes(),
            );
            emit(
                &mut rec,
                b"prec_lower",
                p.lower.map(ieee754_hex).as_deref().unwrap_or("").as_bytes(),
            );
            emit(
                &mut rec,
                b"prec_upper",
                p.upper.map(ieee754_hex).as_deref().unwrap_or("").as_bytes(),
            );
            emit(
                &mut rec,
                b"prec_selected",
                p.selected_mz
                    .map(ieee754_hex)
                    .as_deref()
                    .unwrap_or("")
                    .as_bytes(),
            );
            emit(
                &mut rec,
                b"prec_charge",
                p.charge
                    .map(|c| c.to_string())
                    .as_deref()
                    .unwrap_or("")
                    .as_bytes(),
            );
        }
    }

    emit(
        &mut rec,
        b"array_count",
        arrays.len().to_string().as_bytes(),
    );
    for (role, prec, count, digest) in &arrays {
        emit(&mut rec, b"array_role", role.as_bytes());
        emit(&mut rec, b"array_precision", prec);
        emit(&mut rec, b"array_value_count", count.to_string().as_bytes());
        emit(&mut rec, b"array_hash", digest.as_bytes());
    }

    rec.push(RS);
    Ok(rec)
}

/// Return the sha256 of the canonical content form of an mzML file.
pub fn canonicalize_mzml(path: &Path) -> Result<[u8; 32]> {
    if !path.is_file() {
        return Err(ProvenanceError::MissingArtifact(format!(
            "mzml file not found: {}",
            path.display()
        )));
    }
    let text = std::fs::read_to_string(path)?;
    let doc = Document::parse(&text).map_err(|e| {
        ProvenanceError::MalformedSidecar(format!("mzml file is not valid XML: {e}"))
    })?;
    let root = doc.root_element();

    let mut indexed: Vec<(i64, Node)> = Vec::new();
    for d in root.descendants() {
        if is_ms_elem(&d, "spectrum") {
            let idx: i64 = d
                .attribute("index")
                .ok_or_else(|| {
                    ProvenanceError::MalformedSidecar(
                        "mzml spectrum has missing index attribute".into(),
                    )
                })?
                .parse()
                .map_err(|_| {
                    ProvenanceError::MalformedSidecar(
                        "mzml spectrum has non-integer index attribute".into(),
                    )
                })?;
            indexed.push((idx, d));
        }
    }
    indexed.sort_by_key(|(i, _)| *i);

    let mut hasher = Sha256::new();
    hasher.update(b"TIMSIM-MZML-CANONICAL-v0\x1f");
    for (_, spectrum) in &indexed {
        hasher.update(&spectrum_record(spectrum)?);
    }
    hasher.update(b"\x1fspectrum_count\x1f");
    hasher.update(indexed.len().to_string().as_bytes());
    hasher.update([US]);
    Ok(hasher.finalize().into())
}

/// Compose the mzML content hash: `sha256(MZML_DOMAIN || mzml || US || config)`.
pub fn compose_mzml_content_hash(mzml_hash: &[u8; 32], config_hash: &[u8; 32]) -> [u8; 32] {
    let mut h = Sha256::new();
    h.update(MZML_CONTENT_DOMAIN);
    h.update(mzml_hash);
    h.update([US]);
    h.update(config_hash);
    h.finalize().into()
}

File: implementations/csharp/Mzprov/CanonicalizeMzml.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Xml;
using System.Xml.Linq;

namespace Mzprov;

// Canonical content hashing for mzML. Operates on spectrum content (not
// XML serialization), byte-for-byte compatible with
// spec/canonicalization-mzml-v0.md and the Python reference.
internal static class CanonicalizeMzml
{
    private static readonly XNamespace Ns = "http://psi.hupo.org/ms/mzml";
    private static readonly byte[] US = { 0x1F };
    private static readonly byte[] RS = { 0x1E };
    private static readonly byte[] MzmlContentDomain = Encoding.ASCII.GetBytes("timsim.mzml.v0\x1f");

    private static class Cv
    {
        public const string MsLevel = "MS:1000511";
        public const string ScanStartTime = "MS:1000016";
        public const string PositiveScan = "MS:1000130";
        public const string NegativeScan = "MS:1000129";
        public const string SelectedIonMz = "MS:1000744";
        public const string ChargeState = "MS:1000041";
        public const string IsoTarget = "MS:1000827";
        public const string IsoLowerOff = "MS:1000828";
        public const string IsoUpperOff = "MS:1000829";
        public const string BinaryF64 = "MS:1000523";
        public const string BinaryF32 = "MS:1000521";
        public const string BinaryI64 = "MS:1000522";
        public const string BinaryI32 = "MS:1000519";
        public const string NoCompression = "MS:1000576";
        public const string ZlibCompression = "MS:1000574";
        public const string NumpressLinear = "MS:1002312";
        public const string NumpressPic = "MS:1002313";
        public const string NumpressSlof = "MS:1002314";
        public const string IonMobility = "MS:1002476";
        public const string IonMobilityAlt = "MS:1003006";
    }

    private static readonly HashSet<string> KnownArrayRoleAccessions = new(StringComparer.Ordinal)
    {
        "MS:1000514", "MS:1000515", "MS:1000516", "MS:1000517", "MS:1000595",
        "MS:1000617", "MS:1000786", "MS:1000820", "MS:1000821", "MS:1000822",
        "MS:1002476", "MS:1002477", "MS:1002478", "MS:1003006", "MS:1003007",
        "MS:1003008", "MS:1003153",
    };

    private static readonly HashSet<string> KnownEncodingAccessions = new(StringComparer.Ordinal)
    {
        "MS:1000519", "MS:1000521", "MS:1000522", "MS:1000523", "MS:1000574",
        "MS:1000576", "MS:1002312", "MS:1002313", "MS:1002314",
    };

    public static byte[] Run(string mzmlPath)
    {
        if (!File.Exists(mzmlPath))
        {
            throw new FileNotFoundException($"mzml file not found: {mzmlPath}");
        }

        XDocument doc;
        try
        {
            using var reader = XmlReader.Create(mzmlPath, new XmlReaderSettings { DtdProcessing = DtdProcessing.Prohibit });
            doc = XDocument.Load(reader);
        }
        catch (XmlException e)
        {
            throw new MalformedSidecarException($"mzml file is not valid XML: {e.Message}");
        }

        var root = doc.Root!;
        using var sha = IncrementalHash.CreateHash(HashAlgorithmName.SHA256);
        sha.AppendData(Encoding.ASCII.GetBytes("TIMSIM-MZML-CANONICAL-v0\x1f"));

        int spectrumCount = 0;
        foreach (var spectrum in SpectraSorted(root))
        {
            sha.AppendData(SpectrumRecord(spectrum));
            spectrumCount++;
        }

        sha.AppendData(Encoding.ASCII.GetBytes(
            "\x1fspectrum_count\x1f" + spectrumCount.ToString(CultureInfo.InvariantCulture) + "\x1f"));
        return sha.GetHashAndReset();
    }

    public static byte[] ComposeMzmlContentHash(byte[] mzmlHash, byte[] configHash)
    {
        using var ms = new MemoryStream();
        ms.Write(MzmlContentDomain, 0, MzmlContentDomain.Length);
        ms.Write(mzmlHash, 0, mzmlHash.Length);
        ms.Write(US, 0, US.Length);
        ms.Write(configHash, 0, configHash.Length);
        return SHA256.HashData(ms.ToArray());
    }

    private static IEnumerable<XElement> SpectraSorted(XElement root)
    {
        var indexed = new List<(int Index, XElement Spectrum)>();
        foreach (var s in root.Descendants(Ns + "spectrum"))
        {
            string raw = s.Attribute("index")?.Value ?? "";
            if (!int.TryParse(raw, NumberStyles.Integer, CultureInfo.InvariantCulture, out int idx))
            {
                string sid = s.Attribute("id")?.Value ?? "";
                throw new MalformedSidecarException(
                    $"mzml spectrum '{sid}' has missing or non-integer index");
            }
            indexed.Add((idx, s));
        }
        indexed.Sort((a, b) => a.Index.CompareTo(b.Index));
        return indexed.Select(t => t.Spectrum);
    }

    private static byte[] SpectrumRecord(XElement spectrum)
    {
        string specId = spectrum.Attribute("id")?.Value ?? "";
        string rawIndex = spectrum.Attribute("index")?.Value ?? "";
        if (!int.TryParse(rawIndex, NumberStyles.Integer, CultureInfo.InvariantCulture, out int index))
        {
            throw new MalformedSidecarException(
                $"mzml spectrum '{specId}' has missing or non-integer index attribute");
        }

        string msLevel = CvValue(spectrum, Cv.MsLevel) ?? "";
        byte[] polarity = ExtractPolarity(spectrum);
        double? rt = ExtractRtSeconds(spectrum);
        double? mob = ExtractIonMobility(spectrum);
        var precursor = ExtractPrecursor(spectrum);

        var arrays = new List<(string Role, byte[] Tag, long Count, byte[] Digest)>();
        var bdal = spectrum.Element(Ns + "binaryDataArrayList");
        if (bdal != null)
        {
            foreach (var bda in bdal.Elements(Ns + "binaryDataArray"))
            {
                string roleLabel = ArrayRoleLabel(bda);
                byte[] payload = DecodeBinaryArray(bda);
                byte[] tag = ArrayPrecisionTag(bda);
                int width = PrecisionWidth(tag);
                long count = width != 0 ? payload.Length / width : 0;
                byte[] digest = Encoding.ASCII.GetBytes(Canonicalize.ToHexLower(SHA256.HashData(payload)));
                arrays.Add((roleLabel, tag, count, digest));
            }
        }
        arrays.Sort((a, b) => string.CompareOrdinal(a.Role, b.Role));

        var parts = new List<byte[]>();
        void Emit(string key, byte[] value) =>
            parts.Add(Concat(US, Encoding.ASCII.GetBytes(key), US, value, US));

        Emit("spec_index", Ascii(index.ToString(CultureInfo.InvariantCulture)));
        Emit("spec_id", Encoding.UTF8.GetBytes(specId));
        Emit("ms_level", Ascii(msLevel));
        Emit("polarity", polarity);
        Emit("rt_sec", rt.HasValue ? Ieee754Hex(rt.Value) : Array.Empty<byte>());
        Emit("mobility", mob.HasValue ? Ieee754Hex(mob.Value) : Array.Empty<byte>());

        if (precursor is null)
        {
            Emit("prec_target", Array.Empty<byte>());
            Emit("prec_lower", Array.Empty<byte>());
            Emit("prec_upper", Array.Empty<byte>());
            Emit("prec_selected", Array.Empty<byte>());
            Emit("prec_charge", Array.Empty<byte>());
        }
        else
        {
            Emit("prec_target", precursor.Target.HasValue ? Ieee754Hex(precursor.Target.Value) : Array.Empty<byte>());
            Emit("prec_lower", precursor.Lower.HasValue ? Ieee754Hex(precursor.Lower.Value) : Array.Empty<byte>());
            Emit("prec_upper", precursor.Upper.HasValue ? Ieee754Hex(precursor.Upper.Value) : Array.Empty<byte>());
            Emit("prec_selected", precursor.SelectedMz.HasValue ? Ieee754Hex(precursor.SelectedMz.Value) : Array.Empty<byte>());
            Emit("prec_charge", precursor.Charge.HasValue ? Ascii(precursor.Charge.Value.ToString(CultureInfo.InvariantCulture)) : Array.Empty<byte>());
        }

        Emit("array_count", Ascii(arrays.Count.ToString(CultureInfo.InvariantCulture)));
        foreach (var (role, prec, count, digest) in arrays)
        {
            Emit("array_role", Ascii(role));
            Emit("array_precision", prec);
            Emit("array_value_count", Ascii(count.ToString(CultureInfo.InvariantCulture)));
            Emit("array_hash", digest);
        }

        return Concat(Concat(parts.ToArray()), RS);
    }

    // ----- field extraction -----

    private static byte[] ExtractPolarity(XElement spectrum)
    {
        if (CvPresent(spectrum, Cv.PositiveScan)) return Encoding.ASCII.GetBytes("+");
        if (CvPresent(spectrum, Cv.NegativeScan)) return Encoding.ASCII.GetBytes("-");
        return Encoding.ASCII.GetBytes("?");
    }

    private static double? ExtractRtSeconds(XElement spectrum)
    {
        var scanList = spectrum.Element(Ns + "scanList");
        var scan = scanList?.Element(Ns + "scan");
        if (scan is null) return null;
        foreach (var cv in scan.Elements(Ns + "cvParam"))
        {
            if (cv.Attribute("accession")?.Value == Cv.ScanStartTime)
            {
                if (!TryParseDouble(cv.Attribute("value")?.Value, out double value))
                {
                    return null;
                }
                if (cv.Attribute("unitAccession")?.Value == "UO:0000031") // minute
                {
                    value *= 60.0;
                }
                return value;
            }
        }
        return null;
    }

    private static double? ExtractIonMobility(XElement spectrum)
    {
        foreach (string accession in new[] { Cv.IonMobility, Cv.IonMobilityAlt })
        {
            string? v = CvValueRecursive(spectrum, accession);
            if (v != null && TryParseDouble(v, out double parsed))
            {
                return parsed;
            }
        }
        return null;
    }

    private sealed class Precursor
    {
        public double? Target;
        public double? Lower;
        public double? Upper;
        public double? SelectedMz;
        public int? Charge;

        public bool Any => Target.HasValue || Lower.HasValue || Upper.HasValue ||
                           SelectedMz.HasValue || Charge.HasValue;
    }

    private static Precursor? ExtractPrecursor(XElement spectrum)
    {
        var plist = spectrum.Element(Ns + "precursorList");
        var precursor = plist?.Element(Ns + "precursor");
        if (precursor is null) return null;

        var result = new Precursor();
        var iso = precursor.Element(Ns + "isolationWindow");
        if (iso != null)
        {
            if (TryParseDouble(CvValue(iso, Cv.IsoTarget), out double t)) result.Target = t;
            if (TryParseDouble(CvValue(iso, Cv.IsoLowerOff), out double lo)) result.Lower = lo;
            if (TryParseDouble(CvValue(iso, Cv.IsoUpperOff), out double up)) result.Upper = up;
        }

        var sion = precursor.Element(Ns + "selectedIonList")?.Element(Ns + "selectedIon");
        if (sion != null)
        {
            if (TryParseDouble(CvValue(sion, Cv.SelectedIonMz), out double mz)) result.SelectedMz = mz;
            string? charge = CvValue(sion, Cv.ChargeState);
            if (charge != null && int.TryParse(charge, NumberStyles.Integer, CultureInfo.InvariantCulture, out int c))
            {
                result.Charge = c;
            }
        }

        return result.Any ? result : null;
    }

    // ----- binary array -----

    private static byte[] DecodeBinaryArray(XElement bda)
    {
        foreach (string acc in new[] { Cv.NumpressLinear, Cv.NumpressPic, Cv.NumpressSlof })
        {
            if (CvPresent(bda, acc))
            {
                throw new ProvenanceException(
                    $"mzml binaryDataArray uses numpress compression ({acc}); not supported in canonicalize_mzml v0");
            }
        }

        var binaryEl = bda.Element(Ns + "binary");
        if (binaryEl is null)
        {
            throw new MalformedSidecarException("mzml binaryDataArray is missing the inner <binary> element");
        }
        string text = (binaryEl.Value ?? "").Trim();
        if (text.Length == 0)
        {
            return Array.Empty<byte>();
        }

        byte[] raw;
        try
        {
            raw = Convert.FromBase64String(text);
        }
        catch (FormatException e)
        {
            throw new MalformedSidecarException($"mzml binaryDataArray has undecodable base64 content: {e.Message}");
        }

        if (CvPresent(bda, Cv.ZlibCompression))
        {
            try
            {
                using var input = new MemoryStream(raw);
                using var zlib = new ZLibStream(input, CompressionMode.Decompress);
                using var output = new MemoryStream();
                zlib.CopyTo(output);
                raw = output.ToArray();
            }
            catch (Exception e) when (e is InvalidDataException or IOException)
            {
                throw new MalformedSidecarException($"mzml binaryDataArray has undecompressable zlib payload: {e.Message}");
            }
        }

        return raw;
    }

    private static byte[] ArrayPrecisionTag(XElement bda)
    {
        if (CvPresent(bda, Cv.BinaryF64)) return Encoding.ASCII.GetBytes("f64");
        if (CvPresent(bda, Cv.BinaryF32)) return Encoding.ASCII.GetBytes("f32");
        if (CvPresent(bda, Cv.BinaryI64)) return Encoding.ASCII.GetBytes("i64");
        if (CvPresent(bda, Cv.BinaryI32)) return Encoding.ASCII.GetBytes("i32");
        return Encoding.ASCII.GetBytes("??");
    }

    private static int PrecisionWidth(byte[] tag)
    {
        string t = Encoding.ASCII.GetString(tag);
        if (t == "f64" || t == "i64") return 8;
        if (t == "f32" || t == "i32") return 4;
        return 0;
    }

    private static string ArrayRoleLabel(XElement bda)
    {
        foreach (var cv in bda.Elements(Ns + "cvParam"))
        {
            string acc = cv.Attribute("accession")?.Value ?? "";
            if (KnownArrayRoleAccessions.Contains(acc))
            {
                return acc;
            }
        }
        foreach (var cv in bda.Elements(Ns + "cvParam"))
        {
            string acc = cv.Attribute("accession")?.Value ?? "";
            if (acc.Length > 0 && !KnownEncodingAccessions.Contains(acc))
            {
                return "unknown:" + acc;
            }
        }
        throw new MalformedSidecarException("binaryDataArray has no cvParam identifying its array role");
    }

    // ----- cvParam helpers -----

    private static string? CvValue(XElement element, string accession)
    {
        foreach (var child in element.Elements(Ns + "cvParam"))
        {
            if (child.Attribute("accession")?.Value == accession)
            {
                return child.Attribute("value")?.Value;
            }
        }
        return null;
    }

    private static bool CvPresent(XElement element, string accession)
    {
        foreach (var child in element.Elements(Ns + "cvParam"))
        {
            if (child.Attribute("accession")?.Value == accession)
            {
                return true;
            }
        }
        return false;
    }

    private static string? CvValueRecursive(XElement element, string accession)
    {
        foreach (var child in element.Descendants(Ns + "cvParam"))
        {
            if (child.Attribute("accession")?.Value == accession)
            {
                return child.Attribute("value")?.Value;
            }
        }
        return null;
    }

    // ----- byte helpers -----

    private static byte[] Ieee754Hex(double value)
    {
        byte[] be = BitConverter.GetBytes(value);
        if (BitConverter.IsLittleEndian)
        {
            Array.Reverse(be);
        }
        return Encoding.ASCII.GetBytes(Canonicalize.ToHexLower(be));
    }

    private static bool TryParseDouble(string? s, out double value)
    {
        if (s is null)
        {
            value = 0;
            return false;
        }
        // Mirror Python float(): accept decimal/exponent and inf/nan spellings.
        return double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out value)
            || TryParseSpecialFloat(s, out value);
    }

    private static bool TryParseSpecialFloat(string s, out double value)
    {
        string t = s.Trim().ToLowerInvariant();
        switch (t)
        {
            case "inf":
            case "+inf":
            case "infinity":
            case "+infinity":
                value = double.PositiveInfinity;
                return true;
            case "-inf":
            case "-infinity":
                value = double.NegativeInfinity;
                return true;
            case "nan":
            case "+nan":
            case "-nan":
                value = double.NaN;
                return true;
            default:
                value = 0;
                return false;
        }
    }

    private static byte[] Ascii(string s) => Encoding.ASCII.GetBytes(s);

    private static byte[] Concat(params byte[][] parts)
    {
        int total = 0;
        foreach (var p in parts) total += p.Length;
        var result = new byte[total];
        int offset = 0;
        foreach (var p in parts)
        {
            Array.Copy(p, 0, result, offset, p.Length);
            offset += p.Length;
        }
        return result;
    }
}
